tempesta_fw/connection.h

```diff
@@ -284,6 +289,55 @@ tfw_srv_conn_hasnip(TfwSrvConn *srv_conn)
 	return test_bit(TFW_CONN_B_HASNIP, &srv_conn->flags);
 }
 
+/*
+ * Tell if connection is temporary inactive due to full work queue.
+ */
+static inline bool
+tfw_srv_conn_busy(TfwSrvConn *conn)
+{
+	if (test_bit(TFW_CONN_B_BUSY, &conn->flags))
+		return true;
+	rmb();
+	if (time_is_after_jiffies(READ_ONCE(conn->jbusytstamp)))
+		return true;
+
+	return false;
+}
+
+/*
+ * Make connection temporary inactive due to full work queue.
+ */
+static inline void
+tfw_srv_conn_set_busy(TfwSrvConn *conn)
+{
+	if (!test_bit(TFW_CONN_B_BUSY, &conn->flags))
+		set_bit(TFW_CONN_B_BUSY, &conn->flags);
+}
+
+/*
+ * Clear busy bit for server connection (if set).
+ */
+static inline void
+tfw_srv_conn_clear_busy(TfwSrvConn *conn)
+{
+	if (test_bit(TFW_CONN_B_BUSY, &conn->flags))
+		clear_bit(TFW_CONN_B_BUSY, &conn->flags);
+}
+
+/*
+ * Set small delay for inactivity of busy connection to give time for
+ * unloading of the corresponding work queue.
+ */
+static inline void
+tfw_srv_clear_busy_delay(TfwSrvConn *conn, bool busy)
+{
+	if (busy)
+		WRITE_ONCE(conn->jbusytstamp, jiffies + msecs_to_jiffies(30));
+
+	smp_mb__before_atomic();
```

Comment:
It would be nice to have a comment that explains the logic of memory barriers in this set of functions.

Reply:
Added comment for `tfw_srv_clear_busy_delay()`.